src/server.ts

```diff
@@ -13,107 +12,127 @@ declare interface Context {
+
+  } else { // a basic string or dom element, just get children
+    if (element.props && element.props.children) {
+      Children.forEach(element.props.children, (child: any) => {
+        walkTree(child, context, visitor);
+      });
+    }
   }
+}
+
+function getQueriesFromTree(
+  { rootElement, rootContext = {} }: QueryTreeArgument, fetchRoot: boolean = true
+): QueryResult[] {
+  const queries = [];
+
+  walkTree(rootElement, rootContext, (element, context) => {
+    const Component = element.type || element;
+
+    const skipRoot = !fetchRoot && (element === rootElement);
+    if (typeof Component.fetchData === 'function' && !skipRoot) {
+      const props = assign({}, Component.defaultProps, element.props);
+      const query = Component.fetchData(props, context);
+      if (query) {
+        queries.push({ query, element, context });
+
+        // Tell walkTree to not recurse inside this component;  we will
+        // wait for the query to execute before attempting it.
+        return false;
+      }
+    }
+  });
 
-  return { queries, context: contextStore };
+  return queries;
 }
 
 // XXX component Cache
-export function getDataFromTree(app, ctx: any = {}, fetch: boolean = true): Promise<any> {
+export function getDataFromTree(rootElement, rootContext: any = {}, fetchRoot: boolean = true): Promise<void> {
 
-  // reset for next loop
-  contextStore = {};
-  let { context, queries } = getQueriesFromTree({ component: app, context: ctx }, fetch);
-  // reset for next loop
-  contextStore = {};
+  let queries = getQueriesFromTree({ rootElement, rootContext }, fetchRoot);
 
   // no queries found, nothing to do
-  if (!queries.length) return Promise.resolve(context);
-
-  const mappedQueries = flatten(queries).map(y => y.query.then(x => y));
-  // run through all queries we can
-  return Promise.all(mappedQueries)
-    .then(trees => Promise.all(trees.filter(x => !!x).map((x: any) => {
-      return getDataFromTree(x.component, context, false); // don't rerun `fetchData'
-    })))
-    .then(() => (context));
-
+  if (!queries.length) return Promise.resolve();
+
+  // wait on each query that we found, re-rendering the subtree when it's done
+  const mappedQueries = queries.map(({ query, element, context }) =>  {
+    // we've just grabbed the query for element, so don't try and get it again
+    return query.then(_ => getDataFromTree(element, context, false));
+  });
+  return Promise.all(mappedQueries).then(_ => null);
 }
 
 export function renderToStringWithData(component) {
   return getDataFromTree(component)
-    .then(({ client }) => {
-      let markup = ReactDOM.renderToString(component);
-      let apolloState = client.queryManager.getApolloState();
-
-      for (let queryId in apolloState.queries) {
-        let fieldsToNotShip = ['minimizedQuery', 'minimizedQueryString'];
-        for (let field of fieldsToNotShip) delete apolloState.queries[queryId][field];
-      }
+    .then(() => ReactDOM.renderToString(component));
```

Comment:
Wow, what a massive cleanup! 

Are you planning to keep `initialState` as a return value?